File: src/collider/java/Buf.java
package collider.java;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public final class Buf {
    public byte[] a;
    public int w, r;

    public Buf(int capacity) {
        a = new byte[capacity];
    }

    private void checkRead(int n) {
        if (r + n > w) {
            throw new IndexOutOfBoundsException("read " + n + " at " + r + " of " + w);
        }
    }

    private void need(int n) {
        if (w + n > a.length) {
            byte[] b = new byte[Math.max(a.length * 2, w + n)];
            System.arraycopy(a, 0, b, 0, w);
            a = b;
        }
    }

    public int readableBytes() {
        return w - r;
    }

    public void clear() {
        w = 0;
        r = 0;
    }

    public void clear(int keep) {
        if (a.length > keep) a = new byte[keep];
        w = 0;
        r = 0;
    }

    public void readFrom(InputStream in, int n) throws IOException {
        need(n);
        int got = 0;
        while (got < n) {
            int k = in.read(a, w + got, n - got);
            if (k < 0) throw new IOException("end of stream");
            got += k;
        }
        w += n;
    }

    public void writeTo(OutputStream out) throws IOException {
        out.write(a, r, w - r);
        r = w;
    }

    public void writeByte(int v) {
        need(1);
        a[w++] = (byte) v;
    }

    public void writeBoolean(boolean v) {
        writeByte(v ? 1 : 0);
    }

    public void writeShort(int v) {
        need(2);
        a[w++] = (byte) (v >> 8);
        a[w++] = (byte) v;
    }

    public void writeInt(int v) {
        need(4);
        for (int i = 24; i >= 0; i -= 8) a[w++] = (byte) (v >> i);
    }

    public void writeLong(long v) {
        need(8);
        for (int i = 56; i >= 0; i -= 8) a[w++] = (byte) (v >> i);
    }

    public void writeFloat(float v) {
        writeInt(Float.floatToIntBits(v));
    }

    public void writeDouble(double v) {
        writeLong(Double.doubleToLongBits(v));
    }

    public void writeBytes(byte[] src) {
        writeBytes(src, 0, src.length);
    }

    public void writeBytes(byte[] src, int off, int len) {
        need(len);
        System.arraycopy(src, off, a, w, len);
        w += len;
    }

    public void writeBytes(Buf src) {
        writeBytes(src.a, src.r, src.readableBytes());
        src.r = src.w;
    }

    public byte readByte() {
        checkRead(1);
        return a[r++];
    }

    public boolean readBoolean() {
        checkRead(1);
        return a[r++] != 0;
    }

    public int readUnsignedByte() {
        checkRead(1);
        return a[r++] & 0xFF;
    }

    public short readShort() {
        checkRead(2);
        return (short) (((a[r++] & 0xFF) << 8) | (a[r++] & 0xFF));
    }

    public int readUnsignedShort() {
        return readShort() & 0xFFFF;
    }

    public int readInt() {
        checkRead(4);
        int v = 0;
        for (int i = 0; i < 4; i++) v = (v << 8) | (a[r++] & 0xFF);
        return v;
    }

    public long readLong() {
        checkRead(8);
        long v = 0;
        for (int i = 0; i < 8; i++) v = (v << 8) | (a[r++] & 0xFF);
        return v;
    }

    public float readFloat() {
        return Float.intBitsToFloat(readInt());
    }

    public double readDouble() {
        return Double.longBitsToDouble(readLong());
    }

    public void readBytes(byte[] dst) {
        checkRead(dst.length);
        System.arraycopy(a, r, dst, 0, dst.length);
        r += dst.length;
    }
}

File: src/collider/proto/buf.clj
(ns collider.proto.buf
  "Byte buffers of the wire format."
  (:import (collider.java Buf)
           (java.io ByteArrayInputStream InputStream)
           (java.util.zip Deflater Inflater)))

(set! *warn-on-reflection* true)

(defn buf ^Buf [^long capacity]
  (Buf. (int capacity)))

(defn write-byte! [^Buf b ^long v]
  (.writeByte b (int v)))

(defn write-boolean! [^Buf b v]
  (.writeBoolean b (boolean v)))

(defn write-short! [^Buf b ^long v]
  (.writeShort b (int v)))

(defn write-int! [^Buf b ^long v]
  (.writeInt b (int v)))

(defn write-long! [^Buf b ^long v]
  (.writeLong b v))

(defn write-float! [^Buf b ^double v]
  (.writeFloat b (float v)))

(defn write-double! [^Buf b ^double v]
  (.writeDouble b v))

(defn write-bytes!
  ([^Buf b src]
   (if (instance? Buf src)
     (.writeBytes b ^Buf src)
     (.writeBytes b ^bytes src)))
  ([^Buf b ^bytes src ^long off ^long len]
   (.writeBytes b src (int off) (int len))))

(defn write-utf! [^Buf b ^String s]
  (.writeUtf b s))

(defn read-byte ^long [^Buf b]
  (long (.readByte b)))

(defn read-boolean [^Buf b]
  (.readBoolean b))

(defn read-unsigned-byte ^long [^Buf b]
  (long (.readUnsignedByte b)))

(defn read-short ^long [^Buf b]
  (long (.readShort b)))

(defn read-unsigned-short ^long [^Buf b]
  (long (.readUnsignedShort b)))

(defn read-int ^long [^Buf b]
  (long (.readInt b)))

(defn read-long ^long [^Buf b]
  (.readLong b))

(defn read-float ^double [^Buf b]
  (.readFloat b))

(defn read-double ^double [^Buf b]
  (.readDouble b))

(defn read-bytes! [^Buf b ^bytes dst]
  (.readBytes b dst))

(defn readable-bytes ^long [^Buf b]
  (long (.readableBytes b)))

(defn peek-bytes
  "Copies up to n unread bytes without moving the read point."
  ^bytes [^Buf b ^long n]
  (let [n (min n (long (.readableBytes b)))
        dst (byte-array n)]
    (System/arraycopy (.-a b) (.-r b) dst 0 n)
    dst))

(defn clear!
  "Resets the read and write position.
  The two-arg form also shrinks the backing array down to keep
  bytes when it grew larger."
  ([^Buf b]
   (.clear b))
  ([^Buf b ^long keep]
   (.clear b (int keep))))

(defn ensure!
  "Grows the buffer so n more bytes can be written without resize."
  [^Buf b ^long n]
  (.ensure b (int n)))

(defn adopt!
  "Replaces the backing array with src, read at 0 and write at len."
  [^Buf b ^bytes src ^long len]
  (.adopt b src (int len)))

(defn read-from! [^Buf b in ^long n]
  (.readFrom b in (int n)))

(defn write-to! [^Buf b out]
  (.writeTo b out))

(defn unread-stream
  "Returns a stream over the unread bytes. The read point stays."
  ^InputStream [^Buf b]
  (ByteArrayInputStream. (.-a b) (.-r b) (.readableBytes b)))

(defn leave-unread!
  "Moves the read point so only the last n written bytes stay unread."
  [^Buf b ^long n]
  (set! (.-r b) (int (- (.-w b) n))))

(defn inflate-input!
  "Gives the unread bytes to the inflater. The read point stays."
  [^Buf b ^Inflater i]
  (.setInput i (.-a b) (.-r b) (.readableBytes b)))

(defn deflate-input!
  "Gives the unread bytes to the deflater and marks them read."
  [^Buf b ^Deflater d]
  (.setInput d (.-a b) (.-r b) (.readableBytes b))
  (set! (.-r b) (.-w b)))

(defn deflate!
  "Writes what the deflater gives into the free room of the buffer."
  [^Buf b ^Deflater d]
  (let [a (.-a b)
        w (.-w b)
        k (.deflate d a w (- (alength a) w))]
    (set! (.-w b) (int (+ w k)))))
